app/thermal_management/thermalmgmt.c

```diff
@@ -0,0 +1,237 @@
+ * but rather than changing speed on every single degree change, it is changed
+ * after 8 degrees.
+ */
+#define GET_FAN_SPEED_FOR_TEMP(temp)		(temp & (~0x7))
+
+struct therm_sensor *therm_sensor_tbl;
+struct fan_dev *fan_dev_tbl;
+static u8_t max_adc_sensors;
+static u8_t max_fan_dev;
+static bool thermal_initialized;
+static bool peci_initialized;
+static u8_t fan_duty_cycle[FAN_DEV_TOTAL];
+static bool fan_duty_cycle_change;
+static int cpu_temp;
+/* Using module property, before introducing straps handler */
+static bool fan_override;
+
+static void init_fans(void)
+{
+	int ret;
+
+#ifdef CONFIG_THERMAL_FAN_OVERRIDE
+	fan_override = true;
+	LOG_WRN("Fan SW override enable: %d", fan_override);
+#endif
+
+	/* Get the list of fan devices supported for the board */
+	board_fan_dev_tbl_init(&max_fan_dev, &fan_dev_tbl);
+	LOG_DBG("Board has %d fans", max_fan_dev);
+
+	ret = fan_init(max_fan_dev, fan_dev_tbl);
+	if (ret) {
+		LOG_ERR("Failed to init fan");
+	}
+
+	fan_duty_cycle[FAN_CPU] = CONFIG_THERMAL_FAN_OVERRIDE_VALUE;
+	fan_duty_cycle_change = 1;
+}
+
+static void init_therm_sensors(void)
+{
+	uint8_t adc_ch_bits = 0;
+
+	board_therm_sensor_tbl_init(&max_adc_sensors, &therm_sensor_tbl);
+
+	LOG_INF("Num of thermal sensors: %d", max_adc_sensors);
+
+	for (uint8_t idx = 0; idx < max_adc_sensors; idx++) {
+		LOG_INF("adc ch: %d, for acpi sen: %d",
+			therm_sensor_tbl[idx].adc_ch,
+			therm_sensor_tbl[idx].acpi_loc);
+
+		adc_ch_bits |= BIT(therm_sensor_tbl[idx].adc_ch);
+	}
+
+	LOG_INF("adc ch sensors bit: %x", adc_ch_bits);
+	if (thermal_sensors_init(adc_ch_bits)) {
+		LOG_WRN("Thermal Sensor module init failed!");
+	} else {
+		thermal_initialized = true;
+	}
+
+}
+
+static uint8_t mapped_fan_speed(void)
+{
+	uint8_t cpu_fan_speed;
+
+	switch(cpu_temp) {
+	case 00 ... 19 :
+		cpu_fan_speed = 0;
+		break;
+	case 20 ... 29 :
+		cpu_fan_speed = 20;
+		break;
+	case 30 ... 39:
+		cpu_fan_speed = 30;
+		break;
+	case 40 ... 49:
+		cpu_fan_speed = 50;
+		break;
+	case 50 ... 59:
+		cpu_fan_speed = 75;
+		break;
+	case 60 ... 100:
+	default:
+		cpu_fan_speed = 100;
+		break;
+	}
+
+	return cpu_fan_speed;
+}
+
+static void manage_fan(void)
+{
+	/* EC Self control fan based on CPU thermal info */
+	uint8_t cpu_fan_speed = mapped_fan_speed();
+	LOG_WRN("Fan speed 0x%x", cpu_fan_speed);
+
+	if (fan_duty_cycle[FAN_CPU] != cpu_fan_speed) {
+		fan_duty_cycle[FAN_CPU] = cpu_fan_speed;
+		fan_duty_cycle_change = 1;
+	}
+
+	/* HW/KConfig override takes precedence over every control method
+	 * This is mostly used for PO entry on PO team request
+	 */
+	if (fan_override) {
+		fan_duty_cycle[FAN_CPU] = CONFIG_THERMAL_FAN_OVERRIDE_VALUE;
+		fan_duty_cycle_change = 1;
+	}
+
+	if (fan_duty_cycle_change) {
+		fan_duty_cycle_change = 0;
+
+		for (uint8_t idx = 0; idx < max_fan_dev; idx++) {
+			fan_set_duty_cycle(idx, fan_duty_cycle[idx]);
+		}
+	}
+
+	for (uint8_t idx = 0; idx < max_fan_dev; idx++) {
+		uint16_t rpm;
+
+		fan_read_rpm(idx, &rpm);
```

Comment:
This only demonstrates how to read tach sensor, although it's value not used. Still good to keep it.

Reply:
Yes